### PR TITLE
feat: add upper bound for max epoch in zklogin sig

### PR DESCRIPTION
**Pull request by joyqvq**
## Description 

currently a zklogin sig is valid as long as max_epoch is larger than current epoch. this adds a protocol config that the max_epoch cannot be too large, the config sets the upper bound to 30 + current epoch. 

based on new testing framework in simtest only. 

## Test Plan 

simtest
---
If your changes are not user-facing and do not break anything, you can skip the following section. Otherwise, please briefly describe what has changed under the Release Notes section.

### Type of Change (Check all that apply)

- [x] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [x] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### Release notes
Protocol version upgrades to 42. This adds a feature flag to set the upper bound of the max epoch for a zklogin signature.

### CODE REVIEW

File: crates/sui-types/src/message_envelope.rs [hunk 1]
```diff
@@ -106,7 +106,8 @@ pub trait Message {
 
     /// Verify that the message is from the correct epoch (e.g. for CertifiedCheckpointSummary
     /// we verify that the checkpoint is from the same epoch as the committee signatures).
-    fn verify_epoch(&self, epoch: EpochId) -> SuiResult;
+    fn verify_epoch(&self, epoch: EpochId, upper_bound_for_max_epoch: Option<EpochId>)
```
Comment (lxfind) [inline crates/sui-types/src/message_envelope.rs:109]: Is `upper_bound_for_max_epoch` only used for Transaction? If so I am not convinced it belongs to the general message trait.

File: crates/sui-e2e-tests/tests/multisig_tests.rs [hunk 2]
```diff
@@ -175,8 +178,15 @@ async fn test_multisig_e2e() {
...
 async fn test_multisig_with_zklogin_scenerios() {
-    let test_cluster = TestClusterBuilder::new().with_default_jwks().build().await;
+    let test_cluster = TestClusterBuilder::new()
+        .with_default_jwks()
+        .with_epoch_duration_ms(1000)
+        .build()
+        .await;
     test_cluster.wait_for_authenticator_state_update().await;
+    test_cluster
+        .wait_for_epoch_with_timeout(Some(9), Duration::from_secs(190))
```
Comment (mystenmark) [inline crates/sui-e2e-tests/tests/multisig_tests.rs:188]: why do we need to wait 9 epochs in this test?

File: crates/sui-protocol-config/src/lib.rs [hunk 3]
```diff
@@ -12,7 +12,7 @@ use tracing::{info, warn};
 
 /// The minimum and maximum protocol versions supported by this build.
 const MIN_PROTOCOL_VERSION: u64 = 1;
-const MAX_PROTOCOL_VERSION: u64 = 36;
+const MAX_PROTOCOL_VERSION: u64 = 37;
```
Comment (mystenmark) [inline crates/sui-protocol-config/src/lib.rs:15]: fyi @benr-ml is adding 37 in another PR

File: crates/sui-e2e-tests/tests/multisig_tests.rs (hunk 2 above)
Comment (joyqvq) [reply crates/sui-e2e-tests/tests/multisig_tests.rs:188]: because the zk proof is generated for max epoch = 10.

File: crates/sui-types/src/message_envelope.rs (hunk 1 above)
Comment (joyqvq) [reply crates/sui-types/src/message_envelope.rs:109]: theres actually no concrete type at a few places where the verify_epoch is evoked, i can only access it via the Message trait: 
```
impl<T> Envelope<T, AuthoritySignInfo>
where
    T: Message + AuthenticatedMessage + Serialize,
{
    pub fn verify_signatures_authenticated(
        &self,
        committee: &Committee,
        verify_params: &VerifyParams,
    ) -> SuiResult {
        self.data.verify_epoch(self.auth_sig().epoch, None)?;
```

and 
```
impl<T> Envelope<T, AuthoritySignInfo>
where
    T: Message + UnauthenticatedMessage + Serialize,
{
    pub fn verify_authority_signatures(&self, committee: &Committee) -> SuiResult {
        self.data.verify_epoch(self.auth_sig().epoch, None)?;
```

Comment (joyqvq) [reply crates/sui-types/src/message_envelope.rs:109]: there are 4 concrete types for the message trait, 2 out of them actually just return Ok(()) noop, im guessing its a bit messy to remove the trait?

Comment (lxfind) [reply crates/sui-types/src/message_envelope.rs:109]: I think that just means `verify_epoch` shouldn't be a trait API for `Message`. SenderSignedData and SignedCheckpointSummary should have its own function to check this.

Comment (joyqvq) [reply crates/sui-types/src/message_envelope.rs]: `verify_max_epoch_for_all_sigs` i added this call for SenderSignedData and this is only triggered inside verify_tx. the other other implementations like Checkpoint, TransactionEffects etc should not change its interface

```
                signed_tx.verify_epoch(self.committee.epoch())?;
                signed_tx.verify_max_epoch_for_all_sigs(
                    self.committee.epoch(),
                    self.zk_login_params.zklogin_max_epoch_upper_bound,
                )?;
```

however it 2 iterations of the for loop instead of 1. lmk if there is better idea how to refactor this @lxfind  @mystenmark

File: crates/sui-graphql-rpc/tests/e2e_tests.rs [hunk 4]
```diff
@@ -427,16 +436,37 @@ mod tests {
...
+        // Construct a valid zkLogin transaction data, signature.
+        let (kp, pk_zklogin, inputs) =
+            &load_test_vectors("../sui-types/src/unit_tests/zklogin_test_vectors.json")[1];
```
Comment (joyqvq) [inline crates/sui-graphql-rpc/tests/e2e_tests.rs:445]: i ended up rewriting this test to use a zklogin test vector. due to the max_epoch constraint, the old test vector does work any more (previously setting it to 10, but now if the max_epoch is higher than the upper bound, the signature wont verify)

**Comment by joyqvq** [comment]
i just did another rebase and can i please get a review? 
- for sdk i just updated a test to use the same test case as the zklogin test vector file
- the default upper bound is set to 30 @Jordan-Mysten 
- there is an update to graphql tests, to also use zklogin test vector file @wlmyng  
- @lxfind @mystenmark for the sui core changes and simtests

File: crates/sui-e2e-tests/tests/multisig_tests.rs [hunk 5]
```diff
@@ -738,24 +784,36 @@ async fn test_multisig_legacy_works() {
...
-    let tx = construct_simple_zklogin_multisig_tx(&test_cluster).await;
+    let (tx, legacy_tx) = construct_simple_zklogin_multisig_tx(&test_cluster).await;
+    // feature flag disabled fails latest multisig tx.
     let res = test_cluster.wallet.execute_transaction_may_fail(tx).await;
     assert!(res
         .unwrap_err()
         .to_string()
         .contains("zkLogin sig not supported inside multisig"));
+
+    // legacy multisig fails for the same reason.
```
Comment (mystenmark) [inline crates/sui-e2e-tests/tests/multisig_tests.rs:806]: is this test related to the upper bound feature?

File: crates/sui-types/src/zk_login_authenticator.rs [hunk 6]
```diff
@@ -89,11 +88,32 @@ impl Hash for ZkLoginAuthenticator {
...
+        epoch: EpochId,
+        upper_bound_max_epoch: Option<EpochId>,
+    ) -> SuiResult {
+        // the checks here ensure that `current_epoch + upper_bound_max_epoch >= self.max_epoch >= current_epoch`.
+        // 1. if the config for upper bound is set, ensure that the max epoch in signature is not larger than epoch + upper_bound.
+        if let Some(upper_bound) = upper_bound_max_epoch {
+            if self.get_max_epoch() > epoch + upper_bound {
+                return Err(SuiError::InvalidSignature {
+                    error: format!(
+                        "ZKLogin max epoch too large {}, current epoch {}",
```
Comment (mystenmark) [inline crates/sui-types/src/zk_login_authenticator.rs:102]: The error message should include `upper_bound` either directly, or indirectly, so that the user can figure out what the maximum allowable value is

**Review by mystenmark — APPROVED** [review]
Couple of nits, looks good otherwise

File: crates/sui-types/src/signature.rs [hunk 7]
```diff
@@ -32,6 +32,7 @@ pub struct VerifyParams {
     pub zk_login_env: ZkLoginEnv,
     pub verify_legacy_zklogin_address: bool,
     pub accept_zklogin_in_multisig: bool,
+    pub zklogin_max_epoch_upper_bound: Option<EpochId>,
```
Comment (lxfind) [inline crates/sui-types/src/signature.rs:35]: Should this be Option<u64>? This is not the actual epoch, but just number of epochs of validity? May also want to consider renaming some of these because it's hard to tell which one is actual epoch ID and which one represents a delta (i.e. max number of valid epochs)

File: crates/sui-types/src/signature.rs [hunk 8]
```diff
@@ -41,21 +42,27 @@ impl VerifyParams {
         zk_login_env: ZkLoginEnv,
         verify_legacy_zklogin_address: bool,
         accept_zklogin_in_multisig: bool,
+        zklogin_max_epoch_upper_bound: Option<EpochId>,
```
Comment (lxfind) [inline crates/sui-types/src/signature.rs:45]: ditto

File: crates/sui-types/src/signature.rs [hunk 9]
```diff
@@ -41,21 +42,27 @@ impl VerifyParams {
...
 }
 
 /// A lightweight trait that all members of [enum GenericSignature] implement.
 #[enum_dispatch]
 pub trait AuthenticatorTrait {
-    fn verify_user_authenticator_epoch(&self, epoch: EpochId) -> SuiResult;
+    fn verify_user_authenticator_epoch(
+        &self,
+        epoch: EpochId,
+        upper_bound_max_epoch: Option<EpochId>,
```
Comment (lxfind) [inline crates/sui-types/src/signature.rs:64]: Also this one

**Review by lxfind — COMMENTED** [review]
Overall LGTM.
Please address the type/name confusion.

File: crates/sui-types/src/zk_login_authenticator.rs (hunk 6 above)
Comment (joyqvq) [reply crates/sui-types/src/zk_login_authenticator.rs:102]: good point, revised
